client.go

```diff
@@ -1,137 +1,185 @@
-	header.Add("Content-Type", "application/json")
-	header.Add("Auth-Token", "eyJhbGciOiJIUzI1NiIsImlhdCI6MTY0Njk3ODUyMCwiZXhwIjoxOTYyNDM4NTIwfQ.eyJpZCI6ImM0MmU4ZTA4LWExMDAtMTFlYy1hYTUwLTM2ZjFkMmVmNjY1YyJ9.Am9iOHu2FOpsMME3QiEWJv_xk4sCJYaRAjJvAaATIC8")
-	header.Add("generated_signed_url", "true")
-	return header
-}
+// Dispatch an authorized API GET request
+func (c *IClient) authGet(apiPath string, body io.Reader, header http.Header) (*http.Response, error) {
 
-func makeSearchBody(tag string) ([]byte, error) {
-	filter := SearchFilter{
-		Operator: "AND",
-		Terms: []FilterTerm{{
-			Name:  "metadata._gcvi_tags",
-			Value: tag,
-		}},
-	}
-	schema := SearchCriteriaSchema{
-		DocTypes: []string{"assets"},
-		Filter:   filter,
+	req, err := c.authRequest(http.MethodGet, apiPath, body, header)
+	if err != nil {
+		return nil, err
 	}
-	body, err := json.Marshal(schema)
-	return body, err
+
+	resp, err := c.httpClient.Do(req)
+	return resp, err
 }
 
-func makeSearchRequest(apiUrl string, header http.Header, body []byte) (*http.Request, error) {
-	url, err := url.Parse(apiUrl)
+func (c *IClient) authPost(apiPath string, body io.Reader, header http.Header) (*http.Response, error) {
+
+	req, err := c.authRequest(http.MethodPost, apiPath, body, header)
 	if err != nil {
 		return nil, err
 	}
-	req := &http.Request{
-		Method: http.MethodPost,
-		URL:    url,
-		Header: header,
-		Body:   ioutil.NopCloser(bytes.NewReader(body)),
+
+	resp, err := c.httpClient.Do(req)
+	return resp, err
+}
+
+// Looks for an error message in the response body and parses it into a
+// IError object
+func (c *IClient) parseError(body []byte) error {
+	err := &IError{}
+	if json.Unmarshal(body, err) != nil {
+		return nil
 	}
-	return req, nil
+	return err
 }
 
-func main() {
-	flag.StringVar(&tag, "tag", "GPTeaching", "Tag to search for all GPTeaching assets")
-	client := IconikClient{client: &http.Client{}}
-	header := makeSearchHeader()
-	body, err := makeSearchBody(tag)
+// Attempts to parse a response body into the provided result struct
+func (c *IClient) parseResponse(resp *http.Response, result interface{}) error {
+	defer resp.Body.Close()
+
+	// Read response body
+	body, err := ioutil.ReadAll(resp.Body)
 	if err != nil {
-		fmt.Errorf("makeSearchBody() failed: %v", err)
-		return
+		return err
 	}
-	req, err := makeSearchRequest(searchUrl, header, body)
+
+	if c.Debug {
+		log.Printf("Response: %s", body)
+	}
+
+	// Check response code
+	switch resp.StatusCode {
+	case 200: // Response is OK
+	case 401:
+		if err := c.parseError(body); err != nil {
+			return err
+		}
+		return &IError{
+			Errors: []string{"UNKNOWN; error message not parsable"},
+		}
+	default:
+		if err := c.parseError(body); err != nil {
+			return err
+		}
+		return &IError{
+			Errors: []string{"UNKNOWN; error message not parsable"},
+		}
+	}
+
+	return json.Unmarshal(body, result)
+}
+
+func (c *IClient) makeSearchHeader() http.Header {
```

Comment:
why is this a search header? this seems like a header we need to use for all calls (so bad naming)... but beside, this code shouldn't be a separate call, should be in the NewRequest() call.

Reply:
Ah good point, moved inside the newRequest call